datahub/documents/views.py

```
@@ -0,0 +1,87 @@
+
+        page = self.paginate_queryset(queryset)
+        if page is not None:
+            serializer = self.get_serializer(page, many=True)
+            return self.get_paginated_response(serializer.data)
+
+        serializer = self.get_serializer(queryset, many=True)
+        return Response(serializer.data)
+
+    def create(self, request, *args, **kwargs):
+        """Create and one-time upload URL generation."""
+        response = super().create(request, *args, **kwargs)
+        entity_document = self.get_queryset().get(pk=response.data['id'])
+        response.data['signed_upload_url'] = entity_document.document.get_signed_upload_url()
+
+        return response
+
+    def upload_complete_callback(self, request, *args, **kwargs):
+        """File upload done callback."""
+        model_document = self.get_object()
+
+        if not model_document.document.scan_initiated_on:
+            virus_scan_document.apply_async(
+                args=(
+                    str(model_document.document.pk),
+                    model_document.document.get_signed_url(unsafe=True),
+                )
+            )
+            status = upload_statuses.virus_scanning_scheduled
+        else:
+            status = (
+                upload_statuses.completed
+                if model_document.document.scanned_on
+                else upload_statuses.virus_scanning_in_progress
+            )
+
+        serializer = UploadStatusSerializer(data={'status': status})
+        serializer.is_valid(raise_exception=True)
+        return Response(serializer.data)
+
+    def download(self, request, *args, **kwargs):
+        """Provides download information."""
+        model_document = self.get_object()
+
+        if not model_document.document.scanned_on:
+            raise TemporarilyUnavailableException()
+
+        url = model_document.document.get_signed_url()
+        if not url:
+            return HttpResponseForbidden()
```

Should we raise PermissionDenied (or another API exception) rather than return a response? Seems a bit inconsistent to use raise for one and return for the other.

---

Isn't clear enough that 403 is returned because the file has a virus anyway? Not sure it is.

---

This is indeed inconsistent. Added message to PermissionDenied that the file contains a virus.